README.md
# SpectraScope
Interactive interface to visualize spectra and filter sets for any microscope. Easily customizable.

The idea behind **SpectraScope** is to provide users with an *easy-to-use* web-interface (that can be directly downloaded and run locally) where to visualize fluorophores and microscope filter sets of interest. This can be invaluable when setting up new optical configurations or designing new microscopy-based experiments.

SpectraScope is available both as a webpage and as an ElectronJS-based app. You can choose your flavor and download it from the [releases](https://github.com/ggirelli/SpectraScope/releases) page.

**Notice!** The webpage interface has been developed for GoogleChrome (v74) and FireFox (v66). Thus, it may not be fully compatible with other browsers (e.g., Safari, IExplorer,...).

Live demo
---
A live demo of **SpectraScope** webpage is hosted on github, click on the screenshot below to access it.  
[![demo-screenshot](screenshot.png)](https://ggirelli.github.io/SpectraScope/web/app.html)

Features
---

#### For both webpage and webapp

* Visualize excitation/emission spectra of any fluorophore of interest in the 250-900 nm range. The fluorophore color is automatically selected based on the emission peak in the visible range by using the [wave2color](https://github.com/ggirelli/wave2color) functions.
* Visualize the transmission spectrum of any filter or dicroic mirror of interest, overlayed to the fluorophore spectra.
* Visualize the intensity spectrum of any light source of interest, overlayed to the fluorophore spectra.
* Explore the spectra with the mouse, seeing the precise relative intensity and wavelength of any point of interest.
* Quickly see a brief description of any spectrum by hovering over it with the mouse, and access the raw data by clicking on them.

#### For webapp only

* Manage your own settings template, including microscopes, light sources, optical elements, and fluorophores directly through the interface.

Installation
---

Download your flavor of choice (webpage or webapp) for your Operating System from the [releases](https://github.com/ggirelli/SpectraScope/releases) page.

### Known issues

#### webpage

Please, notice that **GoogleChrome** does not allow `crossorigin` requests from local files (i.e., when the address starts with something like `file://` or `C:/`). Thus, when opening `SpectraScope` locally, we recommend either using Mozilla FireFox or installing this tool on an Apache Server.

Usage
---

1. <u>Click</u> on the sources/filters/mirrors/fluorophores (from the *Input* panel) to add them to the graph.
2. <u>Click</u> on the selected filters/fluorophores (from the *Selected* panel) to remove them from the graph.
3. <u>Hover</u> over a spectrum to visualize a short description below this paragraph.
4. <u>Click</u> on a spectrum to visualize the raw data at the end of the page.

More info are available in our [Wiki](https://github.com/ggirelli/SpectraScope/wiki)!

Contribute
---

We welcome any contributions to `SpectraScope`. Please, refer to the [contribution guidelines](https://github.com/ggirelli/SpectraScope/blob/master/CONTRIBUTING.md) if this is your first time contributing! Also, check out our code of conduct.

License
---

```
MIT License
Copyright (c) 2019 Gabriele Girelli
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased
### ...
- ...

## [0.1.0] - 2019-06-17
### Added
- Implemented as an ElectronJS-based app.

## [0.0.3] - 2019-05-29
### Added
- Few more filters.
- Option to select one microscope at a time.
- Favicon.

### Changed
- Fixed the way some filters were reported.
- Now using d3 v5, with promise returns from `d3.tsv`.
- Split Filters in excitation/emission and dicroic mirrors.

## [0.0.2] - 2019-05-23
### Added
- Light source spectra.
- Prepared basis for SpectraScope modeling.

## [0.0.1] - 2019-05-17

[Unreleased] https://github.com/ggirelli/spectrascope/
[0.0.1] https://github.com/ggirelli/spectrascope/releases/tag/v0.0.1
[0.0.2] https://github.com/ggirelli/spectrascope/releases/tag/v0.0.2
[0.0.3] https://github.com/ggirelli/spectrascope/releases/tag/v0.0.3
[0.1.0] https://github.com/ggirelli/spectrascope/releases/tag/v0.1.0

app/src/js/settings-optelem.js

load_optical_element_list = function() {
	// Load list of optical elements and their data to the settings table
	$("table.optElem-settings-list > tbody").children().remove();

	var selectedTemplate = eset.get("selected-template");
	var opticalElements = eset.get("templates." + selectedTemplate + ".optical_elements");

	var typeList = []

	for (var i = Object.keys(opticalElements).length - 1; i >= 0; i--) {
		var optElemName = Object.keys(opticalElements)[i]
		typeList.push(opticalElements[optElemName].type);
		
		var optElemRow = $("<tr></tr>");
		optElemRow.append($("<th scope='row'>" + optElemName + "</th>"));
		optElemRow.append($("<td>" + opticalElements[optElemName].mid + "</td>"));
		if ( "dm" != opticalElements[optElemName].type ) {
			optElemRow.append($("<td>" + opticalElements[optElemName].width + "</td>"));
		}
		optElemRow.append($("<td>" + opticalElements[optElemName].microscopes.join(", ") + "</td>"));
		optElemRow.append($("<td>" + opticalElements[optElemName].color + "</td>"));
		optElemRow.append($("<td>" + opticalElements[optElemName].details + "</td>"));
		optElemRow.append($("<td>" + opticalElements[optElemName].description + "</td>"));

		var optElemOpts = $("<td></td>");
		optElemOpts.append(
			$("<a href='#'></a>")
				.addClass("btn btn-danger btn-sm mr-2")
				.html("<i class='fas fa-minus' ></i>")
				.attr('data-name', optElemName)
				.click(function(e) {
					e.preventDefault();
					rm_optical_element($(this).attr("data-name"));
				})
				.attr('data-toggle', 'tooltip')
				.attr('data-placement', 'top')
				.attr('title', 'Remove optical element')
				.tooltip()
			);
		optElemOpts.append(
			$("<a href='#'></a>")
				.addClass("btn btn-info btn-sm mr-2")
				.html("<i class='fas fa-file-export' ></i>")
				.attr('data-name', optElemName)
				.click(function(e) {
					e.preventDefault();
					dl_optical_element_spectra($(this).attr("data-name"));
				})
				.attr('data-toggle', 'tooltip')
				.attr('data-placement', 'top')
				.attr('title', 'Download spectra')
				.tooltip()
			);
		optElemRow.append(optElemOpts);

		$("table.optElem-settings-list#settings-" + opticalElements[optElemName].type).append(optElemRow);
	}

	base_types = ["em", "ex", "dm"]
	type_meta = {em:8, dm:7, ex:8}
	for (var i = base_types.length - 1; i >= 0; i--) {
		if ( -1 == typeList.indexOf(base_types[i]) ) {
			$("table.optElem-settings-list#settings-" + base_types[i])
				.append($("<tr><td class='text-center' colspan='" + type_meta[base_types[i]] + "'><i>No elements found.</i></td></tr>"));
		}
	}
}

add_optical_element = function(data) {
	// Add a new optical element
	var selectedTemplate = eset.get("selected-template");
	var required_keys = ['name', 'type', 'mid', 'color', 'details', 'description', 'path'];

	var optElem = {};
	var keys = [];
	for (var i = data.length - 1; i >= 0; i--) {
		var e = data[i];
		keys.push(e.name);
		optElem[e.name] = e.value;
	}

	if ( -1 != keys.indexOf("type") ) {
		if ( "dm" != optElem.type ) required_keys.push("width");
	}

	for (var i = required_keys.length - 1; i >= 0; i--) {
		var k = required_keys[i];
		if ( -1 == keys.indexOf(k) ) {
			toastr.error("Missing required '" + k + "' data. No optical element added.");
			return false;
		}
	}

	optElem.mid = parseFloat(optElem.mid);
	if ( "dm" != optElem.type ) optElem.width = parseFloat(optElem.width);
	optElem.spectra = read_spectra(optElem.path);
	optElem.microscopes = [];

	var optElemName = optElem.name
	delete optElem.name;
	eset.set("templates." + selectedTemplate + ".optical_elements." + optElemName, optElem);

	toastr.success("Optical element '" + optElem.name + "' added.");
	load_optical_element_list();
}

add_optical_element_dialog = function() {
	// Show dialog to add a new optical element
	var addOptElemForm = $("<form class='needs-validation'>\
		<div class='mb-2'>\
			Element name:<input class='form-control' type='text' name='name' required />\
			<small class='invalid-feedback'>An optical element with this name is already present. Try a different one!</small>\
		</div>\
		<div class='mb-2'>\
			Element type:<select class='form-control' name='type' id='type'>\
				<option value='ex' selected>Excitation filter</option>\
				<option value='dm'>Dichroic mirror</option>\
				<option value='em'>Emission filter</option>\
			</select>\
		</div>\
		Element midpoint (nm):<input class='form-control mb-2' type='number' min=250 max=900 name='mid' placeholder='Wavelength at transmission window midpoint/beginning' required />\
		Element width (nm):<input class='form-control mb-2' type='number' min=1 max=1000 name='width' placeholder='Transmission window width' required />\
		<div class='mb-2'>\
			Color:<input class='form-control' type='text' name='color' value='auto' required />\
			<small class='invalid-feedback'>The provided color is not valid, try again!</small>\
			<small>Set to <code>auto</code> to let the script select a color automatically, otherwise provide an hexadec code (e.g., '#FFFFFF')).</small>\
		</div>\
		Details:<input class='form-control mb-2' type='text' name='details' />\
		<div>\
			Description:<input class='form-control mb-2' type='text' name='description' />\
			<small>If left empty, a description will be generated by combining midpoint and width.</small>\
		</div>\
		<div class='mb-2'>\
			Spectrum file:<input class='form-control' type='text' name='path' required />\
			<small class='invalid-feedback'>Something is wrong with the uploaded file. Check the console for more details.</small>\
		</div>\
		<input type='submit' class='btn btn-primary float-right mt-2' />\
		<input type='button' class='btn btn-secondary float-right mr-2 mt-2' value='Cancel' />\
	</form>");

	addOptElemForm.submit(function(e) {
		// Check form content and submit if correct
		e.preventDefault();
		var selectedTemplate = eset.get("selected-template");

		$(this).find(".is-invalid").removeClass("is-invalid");
		$(this).find(".is-valid").removeClass("is-valid");

		var nameElem = $(this).find("input[name='name']");
		var newName = nameElem.val();
		if ( newName == null || 0 == newName.length ) {
			nameElem.addClass("is-invalid");
		} else {
			if ( eset.has("templates." + selectedTemplate + ".optical_elements." + newName) ) {
				nameElem.addClass("is-invalid");
			} else {
				nameElem.addClass("is-valid");
			}
		}

		var midpoint = $(this).find("input[name='mid']");
		var newPeak = midpoint.val();
		var re = /[0-9]+/g;
		if ( re.test(newPeak) ) {
			newPeak = parseInt(newPeak);
			if ( newPeak >= 250 & newPeak <= 900 ) {
				midpoint.addClass("is-valid");
			} else {
				midpoint.addClass("is-invalid");
			}
		} else {
			midpoint.addClass("is-invalid");
		}

		if ( "dm" != $(this).find("select").val() ) {
			var winWidth = $(this).find("input[name='width']");
			var newPeak = winWidth.val();
			var re = /[0-9]+/g;
			if ( re.test(newPeak) ) {
				newPeak = parseInt(newPeak);
				if ( newPeak > 0 & newPeak <= 1000 ) {
					winWidth.addClass("is-valid");
				} else {
					winWidth.addClass("is-invalid");
				}
			} else {
				winWidth.addClass("is-invalid");
			}
		}

		var colorElem = $(this).find("input[name='color']");
		var newColor = colorElem.val();
		if ( newColor == "auto" ) {
			colorElem.addClass("is-valid");
		} else {
			var re = /#?[0-9A-Fa-f]{6}/g;
			if ( re.test(newColor) ) {
				colorElem.addClass("is-valid");
				if ( -1 == newColor.indexOf("#") ) {
					colorElem.val("#" + newColor);
				}
			} else {
				colorElem.addClass("is-invalid");
			}
		}

		var spectraElem = $(this).find("input[name='path']");
		var spectrapath = spectraElem.val();
		if ( check_spectra(spectrapath) ) {
			spectraElem.addClass("is-valid");
		} else {
			spectraElem.addClass("is-invalid");
		}

		if ( 0 == $(this).find(".is-invalid").length ) {
			$(this).addClass("was-validated");

			add_optical_element($(this).serializeArray());

			bootbox.hideAll();
		}
	});

	addOptElemForm.find("select").change(function(e) {
		if ( "dm" == $(this).val() ) {
			$(this).parent().parent().find('input[name="width"]').attr("disabled", true);
		} else {
			$(this).parent().parent().find('input[name="width"]').attr("disabled", false);
		}
	});

	addOptElemForm.find("input[type='button']").click(function(e) {
		// Hide bootbox dialog
		bootbox.hideAll();
		toastr.warning("Nothing done.");
		e.preventDefault();
	});

	addOptElemForm.find("input[name='path']").focus(function(e) {
		// Trigger open file dialog
		$(this).blur();
		var filepath = dialog.showOpenDialog(null, {
			title: "Import spectra file",
			properties: ["openFile", "showHiddenFiles"],
			filters: [
				{ name: "Tabulation-separated spectra", extensions: ['tsv'] }
			]
		});
		if ( typeof filepath !== 'undefined' ) {
			$(this).val(filepath[0]);
			$(this).attr('title', filepath[0]);
		}
		e.preventDefault();
	});

	bootbox.dialog({
		title: 'Add optical element',
		message: addOptElemForm,
		closeButton: false
	});
}

rm_optical_element = function(optElemName) {
	// Remove an optical element. Requires confirmation.
	var selectedTemplate = eset.get("selected-template");

	if ( -1 == Object.keys(eset.get("templates." + selectedTemplate + ".optical_elements")).indexOf(optElemName) ) {
		toastr.error("Optical element '" + optElemName + "' not found.");
		return;
	}

	var msg = "Are you sure you want to cancel the '" + optElemName + "' element?<br/>";
	msg = msg + "<small class='text-danger'>This cannot be undone!</small>";
	bootbox.confirm({
		message: msg,
		callback: function(result){
			if ( result ) {
			    eset.delete("templates." + selectedTemplate + ".optical_elements." + optElemName);
			    toastr.success("The element '" + optElemName + "' has been removed.");
			    load_optical_element_list();
			} else {
				toastr.info("Nothing done.");
			}
		}
	});
}

dl_optical_element_spectra = function(optElemName) {
	// Download optical element spectra
	var selectedTemplate = eset.get("selected-template");

	if ( -1 == Object.keys(eset.get("templates." + selectedTemplate + ".optical_elements")).indexOf(optElemName) ) {
		toastr.error("Element '" + optElemName + "' not found.");
		return;
	}

	write_spectra(eset.get("templates." + selectedTemplate + ".optical_elements." + optElemName + ".spectra"),
		optElemName + ".tsv");
}

$(function() {
	$("#settings-optelem-tab").click(function(e) { load_optical_element_list(); });
	$("#settings-add-optelem-btn").click(function(e) { add_optical_element_dialog(); });
});
